File: 1-sorting/insertion_sort/python/insertion_sort.py


def insertion_sort(numbers, increasing = True):

    for j in range(1, len(numbers)):
        key = numbers[j]
        i = j - 1
        while (i > -1) and (numbers[i] > key if increasing else numbers[i] < key):
            numbers[i + 1] = numbers[i]
            i -= 1
        numbers[i + 1] = key

File: 1-sorting/insertion_sort/c/insertion_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

void insertion_sort(int length, int arr[], bool increasing) {
    int i, j, key;
    for (i = 1; i < length; i++) {
        key = arr[i];
        j = i - 1;
        while (j > -1 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1; 
        }
        arr[j + 1] = key;
    }
    for (int k = 0; k < length; k++) {
        printf("%d ", arr[k]);
    }
    printf("\n");
}

int main() {
    int arr[] = { 1, 5, 4, 2, 6 };
    int length = sizeof(arr) / sizeof(arr[0]);
    insertion_sort(length, arr, true);

    return 0;
}